### PR TITLE
Add support for dependencies and primitive shapes that require them

### PR DESCRIPTION
**Pull request by JordonPhillips**
This adds dependency support and support for timestamps, big ints, and big decimals. The generated test package's types now looks like:

```go
// Code generated by smithy-go-codegen DO NOT EDIT.
package weather

import (
	"time"
)

type CityCoordinates struct {
	Latitude  *float32
	Longitude *float32
}

type CitySummary struct {
	Case   *string
	CityId *string
	Name   *string
	Number *string
}

type GetCityImageInput struct {
	CityId *string
}

type GetCityImageOutput struct {
	Image []byte
}

type GetCityInput struct {
	CityId *string
}

type GetCityOutput struct {
	City        *CitySummary
	Coordinates *CityCoordinates
	Name        *string
}

type GetCurrentTimeOutput struct {
	Time *time.Time
}

type GetForecastInput struct {
	CityId *string
}

type GetForecastOutput struct {
	ChanceOfRain *float32
}

type ListCitiesInput struct {
	NextToken *string
	PageSize  *int32
}

type ListCitiesOutput struct {
	Items     []CitySummary
	NextToken *string
}
```

Additionally, `go mod edit -require` will be called for any dependency outside of the standard library.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoDependency.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.List;
+import software.amazon.smithy.codegen.core.SymbolDependency;
+import software.amazon.smithy.codegen.core.SymbolDependencyContainer;
+
+/**
+ * An enum of all the built-in dependencies used by this package.
+ */
+public enum GoDependency implements SymbolDependencyContainer {
+
+    // The version in the stlib dependencies should reflect the minimum Go version.
```
Comment (kstich) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoDependency.java]: s/stlib/stdlib

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/SymbolVisitor.java [hunk 2]
```diff
@@ -228,15 +235,34 @@ public Symbol memberShape(MemberShape shape) {
...
     public Symbol timestampShape(TimestampShape shape) {
-        // TODO: add once dependency support is in
-        return createSymbolBuilder(shape, "nil").build();
+        return createSymbolBuilder(shape, "*time.Time")
+                .addReference(createNamespaceReference(GoDependency.TIME))
+                .build();
     }
 
     private Symbol.Builder createSymbolBuilder(Shape shape, String typeName) {
-        return Symbol.builder().putProperty("shape", shape).name(typeName);
```
Comment (kstich) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/SymbolVisitor.java]: nit: Why the change? If you want to keep, moving `.putProperty` seems right too.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoWriter.java [hunk 3]
```diff
@@ -37,27 +52,141 @@ public GoWriter(String fullPackageName) {
...
+    public GoWriter addImport(String packageName, String as) {
+        imports.addImport(packageName, as);
+        return this;
+    }
+
+    /**
+     * Adds one or more dependencies to the generated code.
+     *
+     * <p>The dependencies of all writers created by the {@link GoDelegator}
+     * are merged together to eventually generate a package.json file.
```
Comment (kstich) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoWriter.java]: "package.json" seems wrong.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/SymbolVisitor.java (hunk 2 above)
Comment (JordonPhillips) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/SymbolVisitor.java]: That's an artifact of other changes I was considering, but went back on. Looks like I missed putting this back.

**Comment by jasdel** [comment]
I think this update may be missing using pointer types for struct shapes. Without pointer type the SDK won't be able to distinguish if a struct is set or not, e.g. empty struct with no member set, vs struct not set at all.

e..g `GetCityOutput` member `City` should be `*CitySummary`.

**Comment by skmcgrail** [comment]
LGTM, can't actually approve as I don't have permissions yet.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/SymbolVisitor.java [hunk 4]
```diff
@@ -228,15 +235,39 @@ public Symbol memberShape(MemberShape shape) {
...
     @Override
     public Symbol timestampShape(TimestampShape shape) {
-        // TODO: add once dependency support is in
-        return createSymbolBuilder(shape, "nil").build();
+        return createSymbolBuilder(shape, "time.Time")
+                .addReference(createNamespaceReference(GoDependency.TIME))
+                .build();
+    }
+
+    private Symbol.Builder createSymbolBuilder(Shape shape, String typeName, boolean pointable) {
```
Comment (kstich) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/SymbolVisitor.java]: I think this would be cleaner as `createPointableSymbolBuilder(Shape shape, String typeName)` (overloaded with + `String namespace`) that sets the pointable value to true explicitly. This removes the need to drop in a bunch of `false` on non-pointables and indicates the symbols are pointable in the specific visitor methods. This also helps having 3 methods with slightly confusing type signatures: (A, B) + (A, B, C) + (A, B, D).
